### ISSUE TITLE
`errorResult` in the TCK ...

### ISSUE THREAD

**Issue by StrayAlien** [dmn-tck/tck#250]
@agilepro @etirelli @opatrascoiu @tarilabs .

I am sure this has probably been discussed to death.  But I'd like to get some clarification about 'errors' in TCK tests.  I see that many things have 'errorResult' on them and I see variations in the test runners as to what that means.  For the recent PRs I have followed suit to what I believe the TCK is after.  A little blindly ...

Camunda and drools check for an execution result with some kind of error on it, whilst JDMN asserts that if an exception is thrown the test should be marked with errorResult as true.

This is not consistent. And that is just the impls in the TCK.

The spec makes no suggestions about errors beyond built in functions should return null for bad input.  Nothing about halting execution and returning null or throwing an exception.  I am a little concerned that we are even asserting error conditions at all. 

So, happy for input.  I know this is likely a minefield,  but asking anyways.

Greg.

**Comment by opatrascoiu** [comment]
The flag is no part of the DMN standard. It was introduced in the TCK XSD schema some time ago. 

I think the intention was to be able to make the distinction between `null` being returned because it was an error, either syntactic or semantic (e.g. incorrect number of parameters in a function call or incorrect types)  and null values returned by correct expression (e.g. `false and null`).

**Comment by StrayAlien** [comment]
Thanks @opatrascoiu - I do understand it is not DMN.  I guess I am wondering why we are even doing it at all.  Every single test in the TCK that has errorResult="true" expect a final return value of the requested decision to be `null`. It kind of suggests that when you get an error the decision result is null - which I don't believe it necessarily the case.  Like `true or not(123)` will actually result in `true` not `null` despite `not(123)` returning `null`.

I am thinking we should be testing all the cases that something may return null or be 'null-coerced' and assert those .. and not an error result. Functions tests are doing them, and in the null-coercion suite (in PR) I have asserted many others.

But the errorResult thing kind of feels like we're not doing it right.

Comments welcome.

**Comment by agilepro** [comment]
The relevant part of the spec:

_10.3.2.16Error Handling
When a built-in function encounters input that is outside its defined domain, the function SHOULD report or log diagnostic information if appropriate, and SHALL return null._

My understanding is that the function (or expression or whatever has the error) returns null.  Thus if you have a function which is being used as a parameter to another calculation, and that function gets an error it evaluates as null, and the null is passed as the parameter.  

There is no concept of an exception that would stop the flow of control.  Instead, null is passed forward and that might cause other subsequent errors -- or not.

There is also supposed to be a mechanism to report that an error has happened.  There are no standard error messages returned and so we can't test if the right message was delivered, so the TCK tests just check whether there IS an error message.  Note that the error might occur during the processing, but the null might be swallowed by another functions able to handle null,  and so the final result might not be null even though an error occurred.

If the test says that an error is reported, then the test does not pass unless an error is reported.

**Comment by StrayAlien** [comment]
Thanks @agilepro.  My feeling is we test for SHALL and leave SHOULD alone. A impl can be DMN compliant and never note an error. 

We can assert for those situations where a null coercion would happen (like the functions param thing) and in the conjunction/disjunction tests and likely those in 'null-coercion' PR.  So I am not sure I see the validity of testing for an error condition of some sort when an impl does not have to report an error at all.  AFAIK we are not asserting any other SHOULD conditions.

An impl may, for example, choose to treat these situations not as errors, but as 'result is _uncertain_ due to domain data issues'.  After all, FEEL null-coercion is simplistic way of saying 'I am not _certain_ about this result'. (And, as a sidebar, a missed opportunity by DMN spec writers to include 'uncertainty' in decision making - but that is another story).

I think it better to assert the conditions under which null-coercion may occur and leave the notion of 'error' out of it.  

(and I do understand an impl writer can create their own test runner and ignore errorResult totally)

**Comment by opatrascoiu** [comment]
There are two possible use cases:
- `null` is returned because there was an error (e.g. syntactic or semantic)
- `null` is returned because is the correct value (e.g. a decision returns the literal expression `null`)

The error flag was introduced to be able to make a distinction between the above when comparing the actual result and the expected result.

**Comment by agilepro** [comment]
What I remember about this is that we decided to tag test cases that are expected to encounter an error so that a runner MAY check to see that an error was produced.  As was mentioned, there may be a large tree of execution and evaluation, and an error may be hit anywhere along the way.  Execution does not stop with an error, and there could be any number of errors.  So all we have is a boolean saying that at least one error was encountered.

I think this was decision but we did not record this as a decision so we should discuss in the next meeting.

The spec is not entirely clear about exactly what should and should not produce an error.  It is OK for a runner to ignore this error flag.  The result is clear and must be tested, but the error "hint" is useful to help get the correct implementation and to allow vendors to check the error values.  Particularly, test cases that are NOT marked as error should NEVER produce an error for that test.  This negative-error testing is important.

When we merge your tests, there might be cases that are tagged error by implementations, and we will have to discuss adding this flag.

**Comment by tarilabs** [comment]
> test cases that are NOT marked as error should NEVER produce an error for that test

As you originally reported, the DMN specification is very broad in this sense and mark and error should be reported; therefore, imho we don't have a base to enforce this either.

Besides this specific point, I agree with Keith comments generally.

**Comment by agilepro** [comment]
Let me clarify.   If the decision is written, and all the expressions are "correct" in that they are valid expressions with valid inputs and valid results, then in this situation the test should NEVER produce an error in any implementation.

If the expression is "2 + 3" then the answer should be 5 and it should literally never encounter an error.  If it does generate an error while evaluating this, then it is wrong and should be marked a fail.

We know when we are writing tests that are somewhat "on the edge" that require type coercion or in some ways stretch the validity of the expression.  Since we can't define error in black-and-white terms, then we need to treat the flag as a "might be an error".   

Any expression that any implementation might send an error should be marked as "might be an error".

We will argue over exactly which expressions should produce an error, but we can't just pick one implementation as the definition.  We need to consider all implementations, so if ANY implementation considers a test an error, then we should set the error flag.  

By operating in this manner, we end up with something of value: when the flag is not set, there should never be an error, and thus any error that occurs is something that needs to be attended to.

OR .... is it worse than I think?

**Comment by StrayAlien** [comment]
Thanks,

First - no offence intended here and not trying to create a flame war or such.  Just discussing.

I guess I am saying that I don’t feel we need to have any tests for errors except possibly ‘syntax’ or compilation errors, and likely not even those.

If we know there are situations where a null will be the result of something then we should have a test that asserts we get a null for it.

I can see many of the tests in the TCK thus far are not ‘unit’ tests and therefore the actual thing being tested may be buried pretty deep in other stuff.  For tests like those, where any number of things along the execution path could go wrong and cause a null result then _maybe_ this helps narrow it down.

But, with more specific tests (and more of them) we can actually test how things interact and what the expected result is - this is the job of the TCK - to remove interpretation from the DMN.   For example, the matrix of tests for arithmetic and comparison on types is likely to be just enormous.  All the types have to be tested against each other (including null) for each and every operation.  This way, we can assert exactly what we expect to happen in each and every circumstance with no ambiguity and no notion of ‘error’. I think the absence of such tests and the ‘integration’ style of tests so far has made things difficult to pinpoint.

> We know when we are writing tests that are somewhat "on the edge" that require type coercion or in some ways stretch the validity of the expression. Since we can't define error in black-and-white terms, then we need to treat the flag as a "might be an error".

IMO the purpose of this TCK is to make sure the above never happens.   Tests are here to remove ‘might’ from the equation and make things absolute.  I believe we’re here to make sure that validity is not stretched or interpreted but proven or disproven. To make things black and white.  The very fact that we can’t define what ‘error’ actually is makes the notion pretty sniffy.

I do think that a test with an expression like `+)` with a syntax error maybe should be marked as an error but, to counter that I think we’re not into testing a million different ways an error can occur like that - we assume correct model input and that includes correct syntax.

Basically, I think if we’re doing our job here the ‘errorResult’ thing should evaporate.  Out-of domain tests, null-coercion tests or whatever.

**Comment by agilepro** [comment]
We made a decision that the "error" flag is entirely optional and can not be used by the runner for any purpose reliably.  We might change that in the future, but for now:  set the flag if you wish but whether the test is marked or not, the runner should not report anything differently.

See decision number 15:  https://s06.circleweaver.com/weaver/t/wfmc/dmn-reference-implementation/decisionList.htm
